### DISCUSSION TITLE
Experimental workflow orchestration (exploring Bluesky)

### DISCUSSION THREAD

**Discussion by sgbaird** [sparks-baird/self-driving-lab-demo#64] [Experimental workflow orchestration]
@maffettone, what are your thoughts on how an example with [Bluesky](https://github.com/bluesky/bluesky) + SDL-Demo might look to best demonstrate Bluesky? For example, would each of the three LEDs be considered individual pieces of equipment (i.e. the red machine, the blue machine, and the green machine) or make it one device? (or maybe it would be interesting to show how it can be accomplished both ways). The same question can apply to the sensor data, which reads eight distinct wavelengths.

I'm picturing some examples going into a section in [the notebooks directory](https://github.com/sparks-baird/self-driving-lab-demo/blob/main/notebooks/README.md). In terms of interfacing with the demo, [section 1](https://github.com/sparks-baird/self-driving-lab-demo/blob/main/notebooks/README.md#1-blinkt-getting-started) and [section 2](https://github.com/sparks-baird/self-driving-lab-demo/blob/main/notebooks/README.md#2-search-algorithms-using-blinkt) show how to control the hardware and run the algorithms all within the RPi 400 (i.e. no communication between separate devices). Since I've redesigned for the Pico W, I left a snapshot of the code that uses Blinkt! and a full Python implementation at https://github.com/sparks-baird/self-driving-lab-demo/tree/main/src/self_driving_lab_demo_blinkt. [Section 3](https://github.com/sparks-baird/self-driving-lab-demo/blob/main/notebooks/README.md#3-pico-w--micropython-implementation) shows how to interact with a local web server to control the hardware with the Pico W. [Section 4](https://github.com/sparks-baird/self-driving-lab-demo/blob/main/notebooks/README.md#4-controlling-the-pico-w-remotely-iot-style) shows how to interact with it using an IoT-style MQTT interface with the PicoW. [Section 5](https://github.com/sparks-baird/self-driving-lab-demo/blob/main/notebooks/README.md#5-nonwireless-control-of-pico) shows how to send commands and read data via a USB connection on the PicoW.

Excited to see Bluesky in action here!

**Comment by maffettone** [comment]
My instinct would be to approach this by creating 5 ophyd devices. All things in ophyd are called motors and detectors for historical reasons, so bear with me on the semantics:

1. A "motor" device for the red channel  or red LED. (At its core it is a setter with no feedback).
2. A motor device for blue.
3. A motor device for green. 
4. A larger device that houses the red, green, and blue devices. 
5. A detector device that compiles the sensor readings into a single array with rich metadata on read. 

Once I get the hardware running, I think the notebook is the right call to show the basics in section 1, but using the bluesky RunEngine. Then we can re-visit how to dive into the more advanced control loops.

**Reply by sgbaird** [reply]
Great! Thanks for explaining the terminology, and I like the concept of each motor device for the LEDs being housed in a larger device.

**Comment by sgbaird** [comment]
As I'm extending the functionality to a different task https://github.com/sparks-baird/self-driving-lab-demo/discussions/89 and splicing out the task-specific code from the boilerplate code https://github.com/sparks-baird/self-driving-lab-demo/pull/137/files#diff-e3007ea32bf1c21b2ec0d22c19f169646f1fbcbf5609f1b38c5b7a819ff0007c, it's looking more and more like a MicroPython version of experiment orchestration software.

**Comment by sgbaird** [comment]
Some other examples of experiment orchestration software (mostly taken from https://github.com/stars/sgbaird/lists/lab-automation):
- https://github.com/CederGroupHub/alab_management
- https://github.com/Chemios/chemios
- https://github.com/darkreactions/ESCALATE
- https://github.com/AFRL-ARES/ARES_OS
- https://github.com/PALab/place
- https://github.com/PrefectHQ/prefect (general-purpose, but could be powerful with experimental orchestration)

**Comment by sgbaird** [comment]
There is a [workflow orchestration software section](https://github.com/AccelerationConsortium/awesome-self-driving-labs#workflow-orchestration) at https://github.com/AccelerationConsortium/awesome-self-driving-labs
